File: day_6/day_6_1.rb
# frozen_string_literal: true

class TuningTrouble
  def initialize(filename)
    @data = parse_packet_data(filename)
  end

  def run
    puts "The start of packet marker is at character ##{find_start_of_packet}"
  end

  #####
  #
  # Variation 1 -- the fastest variation
  #
  # This version operates the sliding window on the input data, as a byte array,
  # only by passing the positional index to `no_duplicate_characters?` and not ever
  # creating subslices of the sliding windows.
  #
  def parse_packet_data(filename)
    File.read(filename).bytes
  end

  def no_duplicate_characters?(idx)
    !(@data[idx] == @data[idx + 1] ||
       @data[idx] == @data[idx + 2] ||
       @data[idx] == @data[idx + 3] ||
       @data[idx + 1] == @data[idx + 2] ||
       @data[idx + 1] == @data[idx + 3] ||
       @data[idx + 2] == @data[idx + 3])
  end

  def find_start_of_packet
    (@data.size - 4).times do |idx|
      return idx + 4 if no_duplicate_characters?(idx)
    end
    'No packet found'
  end

  #####
  #
  # Variation 2 -- concise, non-ASCII-safe, slow
  #
  # This variation is the most concise UTF-8-safe version. It is also incredibly slow,
  # but it isn't the slowest version!
  #
  # Sacrifice full UTF-8 safely and settle for ASCII only by changing one line:
  #    `@data.bytes.each_cons(4) do |chunk|`
  #
  # def parse_packet_data(filename)
  #   File.read(filename)
  # end
  #
  # def find_start_of_packet
  #   idx = 0
  #   @data.chars.each_cons(4) do |chunk|
  #     return idx + 4 if chunk.uniq == chunk
  #     idx += 1
  #   end
  #
  #   'No packet found'
  # end

  #####
  #
  # Variation 3 -- concise, non-ASCII-safe, slow
  #
  # This variation is nearly as concise as variation 2, but it is much faster.
  #
  # def parse_packet_data(filename)
  #   File.read(filename)
  # end
  #
  # def no_duplicate_characters?(chars)
  #   chars.split('').uniq.size == chars.size
  # end
  #
  # def find_start_of_packet
  #   (@data.size - 4).times do |idx|
  #     return idx + 4 if no_duplicate_characters?(@data[idx, 4])
  #   end
  #   'No packet found'
  # end

  #####
  #
  # Variation 4 -- pretty concise, non-ASCII-safe, slowest
  #
  # This is the slowest version that I have provided in Ruby. It is just brutal.
  #
  # def parse_packet_data(filename)
  #   File.read(filename)
  # end
  #
  # def no_duplicate_characters?(chars)
  #   chars.chars.to_set.size == chars_array.size
  # end
  #
  # def find_start_of_packet
  #   (@data.size - 4).times do |idx|
  #     return idx + 4 if no_duplicate_characters?(@data[idx, 4])
  #   end
  #   'No packet found'
  # end

  #####
  #
  # Variation 5 -- most concise, ASCII-only, pretty fast
  #
  # This is still very concise, but it is substantially faster than variations 2 or 3.
  #
  # def parse_packet_data(filename)
  #   File.read(filename).bytes
  # end
  #
  # def no_duplicate_characters?(chars)
  #   chars.bytes.uniq.size == chars.size
  # end
  #
  # def find_start_of_packet
  #   (@data.size - 4).times do |idx|
  #     return idx + 4 if no_duplicate_characters?(@data[idx, 4])
  #   end
  #   'No packet found'
  # end

  #####
  #
  # Variation 6 -- fastest non-ASCII-safe version
  #
  # This version is the fastest that is safe to use with non-ASCII characters.

  # def parse_packet_data(filename)
  #   File.read(filename)
  # end
  #
  # def no_duplicate_characters?(chars)
  #   !( chars[0] == chars[1] ||
  #      chars[0] == chars[2] ||
  #      chars[0] == chars[3] ||
  #      chars[1] == chars[2] ||
  #      chars[1] == chars[3] ||
  #      chars[2] == chars[3] )
  # end
  #
  # def find_start_of_packet
  #   (@data.size - 4).times do |idx|
  #     return idx + 4 if no_duplicate_characters?(@data[idx, 4])
  #   end
  #   'No packet found'
  # end

  #####
  #
  # Variation 7 -- fastest ASCII-only version
  #
  # The only change is to have parse_packet_data return the data as a byte array.
  # This makes it much faster.
  #
  # def parse_packet_data(filename)
  #   File.read(filename).bytes
  # end
  #
  # def no_duplicate_characters?(chars)
  #   !( chars[0] == chars[1] ||
  #      chars[0] == chars[2] ||
  #      chars[0] == chars[3] ||
  #      chars[1] == chars[2] ||
  #      chars[1] == chars[3] ||
  #      chars[2] == chars[3] )
  # end
  #
  # def find_start_of_packet
  #   (@data.size - 4).times do |idx|
  #     return idx + 4 if no_duplicate_characters?(@data[idx, 4])
  #   end
  #   'No packet found'
  # end
end

TuningTrouble.new(ARGV[0] || 'input.txt').run

File: day_6/day_6_1.rs
// use std::collections::HashSet; // Uncomment if trying the HashSet based solution.
use std::env;

struct TuningTrouble {
    data: String,
}

impl TuningTrouble {
    fn new(filename: &str) -> TuningTrouble {
        let data = Self::parse_packet_data(filename).unwrap();
        TuningTrouble { data }
    }

    fn parse_packet_data(filename: &str) -> Option<String> {
        let data = std::fs::read_to_string(filename).ok()?;
        Some(data)
    }

    fn run(&mut self) {
        println!(
            "The start of packet marker is at character #{}",
            self.find_start_of_packet()
        )
    }

    //###
    //
    // Variation 1 -- concise, ASCII-only, fastest
    //
    // This is the fastest Rust implementation. It is also the most concise. Kinda cool.
    //
    fn find_start_of_packet(&mut self) -> usize {
        self.data
            .as_bytes()
            .windows(4)
            .position(|window| {
                window
                    .iter()
                    .enumerate()
                    .all(|(idx, c)| !window[..idx].contains(c))
            })
            .unwrap()
            + 4
    }

    //###
    //
    // Variation 2 -- almost-the-most-concise, ASCII-only, slowest
    //
    // This is the slowest Rust implementation, but it is also the most concise.
    // It can borrow the `sliding_windows()` method and the `find_start_of_packet()`
    // method from variation 3 to become full-UTF8-safe, but it is still pretty slow,
    // particularly for Rust. Ruby can do full-UTF8-safe faster than this.
    //
    // fn no_duplicate_characters(&self, s: &[u8]) -> bool {
    //     let mut set = HashSet::new();
    //     for c in s {
    //         set.insert(c);
    //     }
    //     set.len() == s.len()
    // }
    //
    // fn find_start_of_packet(&mut self) -> usize {
    //     self.data
    //         .as_bytes()
    //         .windows(4)
    //         .position(|chunk| self.no_duplicate_characters(chunk))
    //         .unwrap()
    //         + 4
    // }

    //###
    //
    // Variation 3 -- very-not-concise, non-ASCII-safe, surprisingly fast
    //
    // It turns out that it is tricky to make a Rust implementation that is
    // safe for the full UTF-8 character set, for the simple reason that Rust
    // doesn't allow _character_ based indexing into strings. `foo[2..4]` returns
    // a _byte_ based slice, not a character slice, and `foo[3]` simply isn't allowed,
    // so it is extra work to build a sliding window of characters over a string.
    // This variation does that, and it is impressively faster than variation 2 because
    // of the use of a simple slice instead of a HashSet to determine if there are
    // duplicate characters.
    //
    // fn sliding_windows<'a>(&self, data: &'a str, size: usize) -> impl Iterator<Item = &'a str> {
    //     data.char_indices().flat_map(move |(from, _)| {
    //         data[from..]
    //             .char_indices()
    //             .skip(size - 1)
    //             .next()
    //             .map(|(to, chr)| &data[from..from + to + chr.len_utf8()])
    //     })
    // }
    //
    // fn no_duplicate_characters(&self, s: &str) -> bool {
    //     let chars = s.chars();
    //     let mut seen = ['\x00', '\x00', '\x00', '\x00'];
    //     let mut step = 0;
    //     for char in chars {
    //         if seen.contains(&char) {
    //             return false;
    //         }
    //         seen[step] = char;
    //         step += 1;
    //     }
    //
    //     true
    // }
    //
    // fn find_start_of_packet(&mut self) -> usize {
    //     let mut idx = 0;
    //     for window in self.sliding_windows(&self.data, 4) {
    //         if self.no_duplicate_characters(window) {
    //             return idx + 4;
    //         }
    //         idx += 1;
    //     }
    //
    //     0
    // }

    //###
    //
    // Variation 4 -- not concise, ASCII-only, third fastest
    //
    // This is the almost the fastest Rust implementation. None of the Rust variations
    // are as concise as what Crystal or Ruby are capable of, but this implementation is
    // very comparable to the second fastest Crystal implementation.
    //
    // fn no_duplicate_characters(&self, bytes: &[u8]) -> bool {
    //     !(bytes[0] == bytes[1]
    //         || bytes[0] == bytes[2]
    //         || bytes[0] == bytes[3]
    //         || bytes[1] == bytes[2]
    //         || bytes[1] == bytes[3]
    //         || bytes[2] == bytes[3])
    // }
    //
    // fn find_start_of_packet(&mut self) -> usize {
    //     self.data
    //         .as_bytes()
    //         .windows(4)
    //         .position(|window| self.no_duplicate_characters(window))
    //         .unwrap()
    //         + 4
    // }

    //###
    //
    // Variation 5 -- not concise, ASCII-only, second fastest
    //
    // This implementation is virtually identical to the fastest Crystal implementation.
    //
    // fn no_duplicate_characters(&self, idx: usize, bytes: &[u8]) -> bool {
    //     let mut seen = [bytes[idx], 0, 0, 0];
    //     let mut step = 1;
    //     while step < 4 {
    //         if seen.contains(&bytes[idx + step]) {
    //             return false;
    //         }
    //         seen[step] = bytes[idx + step];
    //         step += 1;
    //     }
    //
    //     true
    // }
    //
    // fn find_start_of_packet(&mut self) -> usize {
    //     self.data
    //         .as_bytes()
    //         .windows(4)
    //         .position(|window| self.no_duplicate_characters(0, window))
    //         .unwrap()
    //         + 4
    // }
}

fn main() {
    let args: Vec<String> = env::args().collect();
    let mut filename = "input.txt";
    if let Some(arg) = args.get(1) {
        filename = arg;
    }

    TuningTrouble::new(filename).run();
}
